websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java

```diff
@@ -0,0 +1,179 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.websocket;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.eclipse.jetty.ee8.websocket.api.Session;
+import org.eclipse.jetty.ee8.websocket.api.WebSocketListener;
+import org.eclipse.jetty.ee8.websocket.api.WriteCallback;
+import org.eclipse.jetty.ee8.websocket.server.JettyServerUpgradeRequest;
+import org.eclipse.jetty.ee8.websocket.server.JettyServerUpgradeResponse;
+import org.eclipse.jetty.ee8.websocket.server.JettyWebSocketServerContainer;
+import org.kohsuke.MetaInfServices;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+@Restricted(NoExternalUse.class)
+@MetaInfServices(Provider.class)
+public class Jetty12E88Provider implements Provider {
+
+    /**
+     * Number of seconds a WebsocketConnection may stay idle until it expires.
+     * Zero to disable.
+     * This value must be higher than the <code>jenkins.websocket.pingInterval</code>.
+     * Per <a href=https://www.eclipse.org/jetty/documentation/jetty-12/programming-guide/index.html#pg-websocket-session-ping>Jetty 12 documentation</a>
+     * a ping mechanism should keep the websocket active. Therefore, the idle timeout must be higher than the ping
+     * interval to avoid timeout issues.
+     */
+    private static long IDLE_TIMEOUT_SECONDS = Long.getLong("jenkins.websocket.idleTimeout", 60L);
+
+    private static final String ATTR_LISTENER = Jetty12E88Provider.class.getName() + ".listener";
+
+    private boolean initialized = false;
+
+    public Jetty12E88Provider() {
+        JettyWebSocketServerContainer.class.hashCode();
+    }
+
+    private void init(HttpServletRequest req) {
+        if (!initialized) {
+            JettyWebSocketServerContainer.getContainer(req.getServletContext()).setIdleTimeout(Duration.ofSeconds(IDLE_TIMEOUT_SECONDS));
+            initialized = true;
+        }
+    }
+
+    @Override
+    public Handler handle(HttpServletRequest req, HttpServletResponse rsp, Listener listener) throws Exception {
+        init(req);
+        req.setAttribute(ATTR_LISTENER, listener);
+        // TODO Jetty 12 has no obvious equivalent to WebSocketServerFactory.isUpgradeRequest; RFC6455Negotiation?
```

Comment:
This comment is phrased as if it is specific to Jetty 12, but actually it is more general and applies to Jetty 10+ (and may very well apply to Jetty 13, etc in the future). To avoid having to keep updating the comment, why not change it from the particular to the generic:

```suggestion
        // TODO Jetty 10+ has no obvious equivalent to WebSocketServerFactory.isUpgradeRequest; RFC6455Negotiation?
```

That way the comment can remain as-is for years to come, until Jetty develops an obvious equivalent to the old method we used to rely on.

Reply:
but it's part of Jetty12 related class..